dev/archery/archery/cli.py

```diff
@@ -0,0 +1,304 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import click
+from contextlib import contextmanager
+import json
+import logging
+import os
+import re
+from tempfile import mkdtemp, TemporaryDirectory
+
+
+from .benchmark.core import BenchmarkComparator
+from .benchmark.runner import CppBenchmarkRunner
+from .lang.cpp import CppCMakeDefinition, CppConfiguration
+from .utils.cmake import CMakeBuild
+from .utils.git import Git
+from .utils.logger import logger, ctx as log_ctx
+from .utils.source import ArrowSources
+
+
+@click.group()
+@click.option("--debug", type=bool, is_flag=True, default=False,
+              help="Increase logging with debugging output.")
+@click.option("-q", "--quiet", type=bool, is_flag=True, default=False,
+              help="Silence executed commands.")
+@click.pass_context
+def archery(ctx, debug, quiet):
+    """ Apache Arrow developer utilities.
+
+    See sub-commands help with `archery <cmd> --help`.
+
+    """
+    # Ensure ctx.obj exists
+    ctx.ensure_object(dict)
+
+    log_ctx.quiet = quiet
+    if debug:
+        logger.setLevel(logging.DEBUG)
+
+
+def validate_arrow_sources(ctx, param, src):
+    """ Ensure a directory contains Arrow cpp sources. """
+    if isinstance(src, str):
+        if not ArrowSources.valid(src):
+            raise click.BadParameter(f"No Arrow C++ sources found in {src}.")
+        src = ArrowSources(src)
+    return src
+
+
+build_dir_type = click.Path(dir_okay=True, file_okay=False, resolve_path=True)
+# Supported build types
+build_type = click.Choice(["debug", "relwithdebinfo", "release"],
+                          case_sensitive=False)
+# Supported warn levels
+warn_level_type = click.Choice(["everything", "checkin", "production"],
+                               case_sensitive=False)
+
+
+@archery.command(short_help="Initialize an Arrow C++ build")
+@click.option("--src", metavar="<arrow_src>", default=ArrowSources.find(),
+              callback=validate_arrow_sources,
+              help="Specify Arrow source directory")
+# toolchain
+@click.option("--cc", metavar="<compiler>", help="C compiler.")
+@click.option("--cxx", metavar="<compiler>", help="C++ compiler.")
+@click.option("--cxx_flags", help="C++ compiler flags.")
```

Comment:
You use `-` everywhere else, did you want `cxx-flags`?

Reply:
yep.